includes/admin/extensions/class-extensions-api.php

```diff
@@ -0,0 +1,201 @@
+		$option_name = sanitize_key( "edd_extension_{$key}_{$body[ $key ]}_data" );
+		$option      = get_option( $option_name );
+		$is_stale    = ! empty( $option['timeout'] ) && time() > $option['timeout'];
+
+		// The ProductData class.
+		$product_data = new ProductData();
+
+		// If the data is "fresh" and what we want exists, return it.
+		if ( ! $is_stale ) {
+			if ( $item_id && ! empty( $option[ $item_id ] ) ) {
+				return $product_data->fromArray( $option[ $item_id ] );
+			} elseif ( ! empty( $option['timeout'] ) ) {
+				unset( $option['timeout'] );
+
+				return $option;
+			}
+		}
+
+		// Get all of the product data.
+		$all_product_data = $this->get_all_product_data();
+
+		// If no product data was retrieved, let the option sit for an hour.
+		if ( ! $all_product_data ) {
+			$data = array(
+				'timeout' => strtotime( '+1 hour', time() ),
+			);
+			if ( $option && $is_stale ) {
+				$data = array_merge( $option, $data );
+			}
+			update_option(
+				$option_name,
+				$data,
+				false
+			);
+
+			if ( $item_id && ! empty( $option[ $item_id ] ) ) {
+				return $product_data->fromArray( $option[ $item_id ] );
+			}
+			unset( $option['timeout'] );
+
+			return $option;
+		}
+
+		$value = array(
+			'timeout' => strtotime( '+1 week', time() ),
+		);
+		if ( $item_id && ! empty( $all_product_data->$item_id ) ) {
+			$item              = $all_product_data->$item_id;
+			$value[ $item_id ] = $this->get_item_data( $item );
+		} elseif ( in_array( $key, array( 'category', 'tag' ), true ) ) {
+			$term_id = $body[ $key ];
+			foreach ( $all_product_data as $item_id => $item ) {
+				if ( 'category' === $key && ( empty( $item->categories ) || ! in_array( $term_id, $item->categories, true ) ) ) {
+					continue;
+				} elseif ( 'tag' === $key && ( empty( $item->tags ) || ! in_array( $term_id, $item->tags, true ) ) ) {
+					continue;
+				}
+				$value[ $item_id ] = $this->get_item_data( $item );
+			}
+		}
+
+		update_option( $option_name, $value, false );
+		unset( $value['timeout'] );
+
+		return $item_id && ! empty( $value[ $item_id ] ) ? $product_data->fromArray( $value[ $item_id ] ) : $value;
+	}
+
+	/**
+	 * Gets all of the product data, either from a tranient or an API request.
+	 *
+	 * @since 2.11.x
+	 * @return object|false
+	 */
+	private function get_all_product_data() {
+		// Possibly all product data is in a transient. If it is, return it.
+		$all_product_data = get_transient( 'edd_all_extension_data' );
+		if ( $all_product_data ) {
+			return $all_product_data;
+		}
+
+		// Otherwise, query the API.
+		$url     = add_query_arg(
+			array(
+				'edd_action' => 'extension_data',
+			),
+			$this->get_products_url()
+		);
+		$request = wp_remote_get(
+			$url,
+			array(
+				'timeout'   => 15,
+				'sslverify' => true,
+			)
+		);
+
+		// If there was an API error, set timeout for 1 hour and return false.
+		if ( is_wp_error( $request ) || ( 200 !== wp_remote_retrieve_response_code( $request ) ) ) {
+			return false;
```

Comment:
Wondering if we should still cache this for 1 hour.

The individual products will be cached, but if this fails, then each product will still try again. So if you delete all your options, make sure the transient isn't set, then visit the necessary pages to trigger each of the 4 promos, that will actually make 4 API requests. Once the 4 are made, it won't make any more, but I still feel like we only need to be making 1.

Reply:
It is cached for an hour--not here, but in lines 48-59. So a failed request should happen for the first product, and then products 2-4 would use the stale data, or no data, depending on what was stored for the option.

Reply:
The use case I'm looking at is if there's no stale data at all and the API request fails. In my tests it's then doing 5 API requests. Logic is like this:

- Email Marketing goes "give me product data". There's no saved option, so it calls `get_all_product_data()` and does an API request. This fails, so marketing data is saved as empty for 1 hour. No more marketing API requests.
- Invoices goes "give me product data". There's no saved option for it, so it calls `get_all_product_data()` and does an API request. This fails, so invoices data is saved as empty for 1 hour. No more Invoices-triggered API requests.
- Rinse and repeat for each extension.

So each individual one does get cached for 1 hour, but each extension _individually_ triggers another call to `get_all_product_data()` if their option doesn't already exist.

To test:

1. Delete transient.
2. Delete all options.
3. Right here https://github.com/awesomemotive/easy-digital-downloads/blob/3a97fae5867a1b5fd195ad00cfc73c3684025e79/includes/admin/extensions/class-extensions-api.php#L120 add these lines:

```
error_log('did API request');
$request = new \WP_Error();
```

Then navigate to pages that would trigger this and you'll see it trigger 5 times (assuming all extensions it's advertising are not active).

Reply:
Okay, see what you think now--I've set it to set the hour long transient as an empty array if the request fails, and updated the conditional checks accordingly (because `get_transient` will return `false` if it doesn't exist).

Reply:
Works great now. :raised_hands: